### PR TITLE
Retoot add accts closes #155

### LINKED ISSUE #155: Replying to toot on another server fails to add the domain to the user handle
When you reply to a toot that originated on another server it does not add the server address to the handle. You  end up with `@` + `username` rather than `@` +`acct`. `acct` includes the domains, eg `@user` vs `@user@domain` . This is also an issue when you want to include more people in a response.  

## EDIT
Each status has a `mentions` which could automatically be pre-populated in a response toot when responding. Note there are `mentions` in both the status and the boosted status. They should be the same, but for ease of use we can put both in a list and insert a unique set into the reply.

### PR DESCRIPTION
**Pull request by alexjgriffith**
## Description
When responding to toots the full acct for both local and federated accounts are now added to the new toot buffer.

## Changes
- Added a function in mastodon.el to return the current user acct
- Added mastodon-toot--process-local, which takes an acct and appends the current server if it is local returns an empty string if the acct matches the current user and does only adds a prefix @ if the acct is federated
- mastodon-toot--mentions will return a formatted string of mentions or an empty string
- adds tests for mastodon-toot--mentions

## Edit (more changes)
- adds a missing , in mastodon-http--post
- `mastodon-toot--reply` now passes `mastodon-toot` a toot-id rather than the whole json
- 'mastodon-toot--reply-to-id is now a local var in a new toot
- only removes one trailing char rather than two from content in toot

### CODE REVIEW

File: lisp/mastodon-toot.el [hunk 1]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
...
   (let* ((toot (mastodon-tl--property 'toot-json))
          (id (mastodon-tl--as-string (mastodon-tl--field 'id toot)))
          (account (mastodon-tl--field 'account toot))
-         (user (cdr (assoc 'username account))))
-    (mastodon-toot user id)))
+         (user (cdr (assoc 'acct account)))
+         (mentions (mastodon-toot--mentions toot)))
+    (mastodon-toot (when user (concat (mastodon-toot--process-local user)
+                                      mentions))
+                   toot)))
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: That second arg is meant to be an id not the whole toot.

File: lisp/mastodon-auth.el [hunk 2]
```diff
@@ -43,6 +43,9 @@
 (defvar mastodon-auth--token-alist nil
   "Alist of User access tokens keyed by instance url.")
 
+(defvar mastodon-auth--acct-alist nil
+  "Alist of acct names keyed by instance url.")
```
Comment (hdurer) [inline lisp/mastodon-auth.el]: Is it just the account names or the whole alist of account details?

File: lisp/mastodon-toot.el [hunk 3]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
       (mastodon-http--triage response
                              (lambda () (message "Toot toot!"))))))
 
+(defun mastodon-toot--get-acct()
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: Unused function.

File: lisp/mastodon-toot.el [hunk 4]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
...
+               (acct (concat "@" username "@" domain)))
+          (kill-new acct)
+          (message (format "Added %s to the kill ring." acct))
+          t)
+      (progn (message (format "No properly formated acct at point:%s." (point)))
+             nil))))
+
+(defun mastodon-toot--process-local (acct)
+  "Adds domain to local ACCT and replaces the curent user name with \"\"."
+  (cond ((string-match-p "@" acct) (concat "@" acct " "))
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: Can you add comments, which cases you are dealing with here?

Also: why are we adding a white space?

File: lisp/mastodon-toot.el [hunk 5]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
...
+               (username (string-remove-prefix "@" (car (last split-address))))
+               (acct (concat "@" username "@" domain)))
+          (kill-new acct)
+          (message (format "Added %s to the kill ring." acct))
+          t)
+      (progn (message (format "No properly formated acct at point:%s." (point)))
+             nil))))
+
+(defun mastodon-toot--process-local (acct)
+  "Adds domain to local ACCT and replaces the curent user name with \"\"."
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: I don't understand this comment.

Does `acct` have to be a local name?  What is the "current user name"?

File: lisp/mastodon-toot.el [hunk 6]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
...
+  (cond ((string-match-p "@" acct) (concat "@" acct " "))
+        ((string-match-p (mastodon-auth--user-acct) acct) "")
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: Ah, is this removing the current user?

That seems orthogonal to what this function is meant to do.

Remove this and explicitly remove the current user from the list you produce with this?

Comment (hdurer) [reply lisp/mastodon-toot.el]: Also: use `string=` rather than `string-match-p`.

File: lisp/mastodon-toot.el [hunk 7]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
...
+        (t (concat "@" acct "@"
+                   (cadr(split-string mastodon-instance-url "/" t)) " "))))
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: nit: space after `cadr`.

File: lisp/mastodon-toot.el [hunk 8]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
...
+
+(defun mastodon-toot--mentions (status)
+  "Extract mentions from STATUS and process them into a string."
+  (interactive)
+  (let ((mentions (cdr(assoc 'mentions status))))
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: nit: space after `cdr`

File: lisp/mastodon-toot.el [hunk 9]
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
...
+    (if (> (length mentions) 0)
+        (mapconcat (lambda(x) (mastodon-toot--process-local
+                               (cdr(assoc 'acct x))))
+                   ;; reverse does not work on vectors in 24.5
+                   (reverse (mapcar (lambda(x) x) mentions)) "") "")))
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: - There is a function `identity` which could replace your lambda
- The [manual](https://www.gnu.org/software/emacs/manual/html_node/elisp/Vector-Functions.html) mentions `(append avector nil)` as a way to convert a vector to a list.

Comment (hdurer) [inline lisp/mastodon-toot.el]: Put that third arg to `mapconcat` on its own line to make it clearer.

Also: that should probably be a space `" "` so that names are properly separated and you don't have to add the space in `mastodon-toot--process-local`.

File: test/mastodon-toot-tests.el [hunk 10]
```diff
@@ -1,6 +1,46 @@
...
+               ((id . "1")
+                (username . "local")
+                (url . "")
+                (acct . "local"))])))
+
+(defconst mastodon-toot-no-mention
+  '((mentions . [])))
+
+(ert-deftest toot-multi-mentions ()
+  (should (string= (let ((mastodon-auth--acct-alist '(("https://local.social". "null")))
```
Comment (hdurer) [inline test/mastodon-toot-tests.el]: I think this is more readable if you pull the `let` out.
That's basically the setup of the test context and now part of the actual test.

File: lisp/mastodon-toot.el (hunk 9 above)
Comment (alexjgriffith) [reply lisp/mastodon-toot.el]: I add the spaces in `mastodon-toot-process local` to save us having to remove repeated spaces introduced from the null case (where your username is in the reply list).

Comment (alexjgriffith) [reply lisp/mastodon-toot.el]: Perfect, was looking for a way to do the conversion cleanly, this seemed unnecessary.

File: lisp/mastodon-auth.el (hunk 2 above)
Comment (alexjgriffith) [reply lisp/mastodon-auth.el]: Just the account names

File: lisp/mastodon-toot.el (hunk 3 above)
Comment (alexjgriffith) [reply lisp/mastodon-toot.el]: This function is meant to be used interactively by the user. Is there a specific way to mark such a function?

File: lisp/mastodon-toot.el (hunk 4 above)
Comment (alexjgriffith) [reply lisp/mastodon-toot.el]: I add the white space here to account for the nil case. In mapconcat an extra space will be added.

File: lisp/mastodon-auth.el [hunk 11]
```diff
@@ -79,5 +82,20 @@ Generate token and set if none known yet."
         (push (cons mastodon-instance-url token) mastodon-auth--token-alist)))
     token))
 
+(defun mastodon-auth--get-account ()
+  "Request user credentials and return an acct."
```
Comment (hdurer) [inline lisp/mastodon-auth.el]: Nit "account name"  Maybe even make that clear in the function name?

File: lisp/mastodon-tl.el [hunk 12]
```diff
@@ -318,7 +318,7 @@ also render the html"
    (concat
     (mastodon-tl--spoiler toot)
     ;; remove two trailing newlines
-    (substring (mastodon-tl--content toot) 0 -2)
+    (substring (mastodon-tl--content toot) 0 -1)
```
Comment (hdurer) [inline lisp/mastodon-tl.el]: That comment above talks about 2 spaces being removed...

File: lisp/mastodon-toot.el [hunk 13]
```diff
@@ -142,14 +141,61 @@ Set `mastodon-toot--content-warning' to nil."
...
+(defun mastodon-toot--process-local (acct)
+  "Adds domain to local ACCT and replaces the curent user name with \"\".
+
+Mastodon requires the full user@domain, even in the case of local accts. 
+eg. \"user\" -> \"user@local.social \" (when local.social is the domain of the 
+mastodon-instance-url).
+eg. \"yourusername\" -> \"\"
+eg. \"feduser@fed.social\" -> \"feduser@fed.social\""
+  (cond ((string-match-p "@" acct) (concat "@" acct " ")) ; federated acct
+        ((string-match-p (mastodon-auth--user-acct) acct) "") ; your acct
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: Again: why not `string=` ?  Is there extra whitespace somewhere to be ignored?

File: lisp/mastodon-toot.el [hunk 14]
```diff
@@ -142,14 +141,61 @@ Set `mastodon-toot--content-warning' to nil."
...
+        (t (concat "@" acct "@" ; local acct
+                   (cadr (split-string mastodon-instance-url "/" t)) " "))))
+
+(defun mastodon-toot--mentions (status)
+  "Extract mentions from STATUS and process them into a string."
+  (interactive)
+  (let ((mentions (cdr (assoc 'mentions status))))
+    (if (> (length mentions) 0)
```
Comment (hdurer) [inline lisp/mastodon-toot.el]: Nit: This check isn't needed, right?  If mentions is the empty vector, the mapappend will still do the right thing.

File: test/mastodon-toot-tests.el [hunk 15]
```diff
@@ -1,6 +1,47 @@
...
+  (let ((mastodon-auth--acct-alist
+         '(("https://local.social". "null")))
+        (mastodon-instance-url "https://local.social"))    
+    (should (string= (mastodon-toot--mentions mastodon-toot-no-mention) ""))))
+
 (ert-deftest cancel ()
   (with-mock
     (mock (kill-buffer-and-window))
     (should (eq nil (mastodon-toot--cancel)))))
+
```
Comment (hdurer) [inline test/mastodon-toot-tests.el]: Nit: remove the empty lines at end?

**Review by hdurer — COMMENTED** [review]
Great.  Only a few nitty comments.

**Comment by hdurer** [comment]
Oh.  About the function to copy the account name:  why not add it to the keymap so users have easy access to it (and document it in the README file)?

**Review by jdenen — APPROVED** [review]
